### PR TITLE
[libc++] Use __is_pointer_in_range for char_traits checks

### PR DESCRIPTION
**Pull request by philnik777**
This allows us to also check the constraints during constant evaluation.

### CODE REVIEW

File: libcxx/include/__string/char_traits.h [hunk 1]
```diff
@@ -142,14 +143,12 @@ struct _LIBCPP_DEPRECATED_("char_traits<T> for T not equal to char, wchar_t, cha
...
-        if (!__libcpp_is_constant_evaluated()) {
-            _LIBCPP_ASSERT_NON_OVERLAPPING_RANGES(
-                __s2 < __s1 || __s2 >= __s1 + __n, "char_traits::copy overlapped range");
-        }
-        char_type* __r = __s1;
-        for (; __n; --__n, ++__s1, ++__s2)
-            assign(*__s1, *__s2);
-        return __r;
+      _LIBCPP_ASSERT_NON_OVERLAPPING_RANGES(
+          !std::__is_pointer_in_range(__s1, __s1 + __n, __s2), "char_traits::copy overlapped range");
```
Comment (ldionne) [inline libcxx/include/__string/char_traits.h:147]:
```suggestion
          !std::__is_pointer_in_range(__s1, __s1 + __n, __s2), "char_traits::copy: source and destination ranges overlap");
```

Everywhere

**Review by ldionne — APPROVED** [review]

**Comment by bgra8** [comment]
heads-up @philnik777: we have about 30 test failures that have all bisected to this revision.

I noticed `std::__is_pointer_in_range()` has a deliberate UB. Replacing that with the `std::reinterpret_cast<const char*>` implementation and then rebuilding the compiler and `libc++` makes all tests pass.

Should we revert this?

**Comment by philnik777** [comment]
> heads-up @philnik777: we have about 30 test failures that have all bisected to this revision.
> 
> I noticed `std::__is_pointer_in_range()` has a deliberate UB. Replacing that with the `std::reinterpret_cast<const char*>` implementation and then rebuilding the compiler and `libc++` makes all tests pass.
> 
> Should we revert this?

I'd like to understand the issue first. What deliberate UB are you talking about?

**Comment by bgra8** [comment]
Not at my desk now. The UB is documented in the implementation
of  __is_pointer_in_range() func.

**Comment by philnik777** [comment]
If you mean the `// Checking this for unrelated pointers is technically UB, but no compiler optimizes based on it (currently).` - that's also the case without this change, it's just not explicitly documented.

**Pull request merged**